File: other/07/puzzle07.py

def minimize_fuel(positions):
    min_position = min(positions)
    max_position = max(positions)
    #print(f"searching from {min_position} to {max_position}")
    best_total_fuel = 2**62
    best_position = None
    for possible_position in range(min_position, max_position+1):
        total_fuel = sum([abs(x-possible_position) for x in positions])
        #print(f"total fuel to move to position {possible_position}: {total_fuel}")
        if total_fuel < best_total_fuel:
            best_total_fuel = total_fuel
            best_position = possible_position

    return best_position, best_total_fuel


SAMPLE_INPUT = "16,1,2,0,4,2,7,1,2,14"

def sample_part1():
    print("SAMPLE Part 1")
    positions = [int(x) for x in SAMPLE_INPUT.split(",")]
    best_position, total_fuel = minimize_fuel(positions)
    print(f"best position = {best_position}, total_fuel = {total_fuel}")
    assert best_position == 2
    assert total_fuel == 37


def part1(positions):
    print("Part 1")
    best_position, total_fuel = minimize_fuel(positions)
    print(f"best position = {best_position}, (p1 answer) total_fuel = {total_fuel}") # 


if __name__ == "__main__":
    input07 = open("input07", encoding="utf-8").read().strip()
    initial_positions = [int(x) for x in input07.split(",")]

    sample_part1()
    part1(initial_positions)

File: aoc2021kotlin/src/main/kotlin/Puzzle07.kt
package net.blergh.advent2021

import java.io.File

object Puzzle07 {
    fun run() {
        val input07 = File("${Main.aocRoot}/other/07/input07").readText().trim()
        val initialPositions = input07.split(",").map { it.toInt() }

        samplePart1()
        part1(initialPositions)
        samplePart2()
        part2(initialPositions)
    }

    private fun minimizeFuelPart1(positions: List<Int>): Pair<Int, Int> {
        val low = positions.minOf { it }
        val high = positions.maxOf { it }
        var bestFuel = Int.MAX_VALUE
        var bestPosition = -1
        for(p in low..high) {
            val totalFuel = positions.map { kotlin.math.abs(it - p) }.sum()
            if(totalFuel < bestFuel) {
                bestPosition = p
                bestFuel = totalFuel
            }
        }

        return Pair(bestPosition, bestFuel)
    }

    private fun minimizeFuelPart2(positions: List<Int>): Pair<Int, Int> {
        val low = positions.minOf { it }
        val high = positions.maxOf { it }
        var bestFuel = Int.MAX_VALUE
        var bestPosition = -1
        for(p in low..high) {
            val totalFuel = positions.map {
                val distance = kotlin.math.abs(it - p)
                (distance * (distance + 1)) / 2
            }.sum()
            if(totalFuel < bestFuel) {
                bestPosition = p
                bestFuel = totalFuel
            }
        }

        return Pair(bestPosition, bestFuel)
    }

    private fun part1(positions: List<Int>) {
        println("Part 1")

        val (bestPosition, totalFuel) = minimizeFuelPart1(positions)
        println("least fuel used at position $bestPosition: $totalFuel") // p=339, fuel=343468
    }

    private fun part2(positions: List<Int>) {
        println("Part 2")

        val (bestPosition, totalFuel) = minimizeFuelPart2(positions)
        println("least fuel used at position $bestPosition: $totalFuel") // p=478, fuel=96086265
    }

    private fun samplePart1() {
        println("SAMPLE P1")

        val positions = SAMPLE_INPUT.split(",").map { it.toInt() }
        val (bestPosition, totalFuel) = minimizeFuelPart1(positions)

        println("least fuel used at position $bestPosition: $totalFuel")
        assert(bestPosition == 2)
        assert(totalFuel == 37)
    }

    private fun samplePart2() {
        println("SAMPLE P2")

        val positions = SAMPLE_INPUT.split(",").map { it.toInt() }
        val (bestPosition, totalFuel) = minimizeFuelPart2(positions)

        println("least fuel used at position $bestPosition: $totalFuel")
        assert(bestPosition == 5)
        assert(totalFuel == 168)
    }


    private const val SAMPLE_INPUT = "16,1,2,0,4,2,7,1,2,14"
}